### ISSUE TITLE
ForwardDiff asserts while precompiling

### ISSUE THREAD

**Issue by Keno** [JuliaLang/julia#22164]
```
julia> using ForwardDiff

julia-debug: /home/keno/julia/src/gf.c:1122: check_ambiguous_visitor: Assertion `!jl_subtype((jl_value_t*)sig, (jl_value_t*)type)' failed.

signal (6): Aborted
while loading /home/keno/.julia/v0.7/ForwardDiff/src/dual.jl, in expression starting on line 195
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x7fbe6dc73bb6)
__assert_fail at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
check_ambiguous_visitor at /home/keno/julia/src/gf.c:1122
jl_typemap_intersection_node_visitor at /home/keno/julia/src/typemap.c:488
jl_typemap_intersection_visitor at /home/keno/julia/src/typemap.c:545
jl_typemap_intersection_visitor at /home/keno/julia/src/typemap.c:536
check_ambiguous_matches at /home/keno/julia/src/gf.c:1209
jl_method_table_insert at /home/keno/julia/src/gf.c:1395
jl_method_def at /home/keno/julia/src/method.c:737
eval at /home/keno/julia/src/interpreter.c:303
eval_body at /home/keno/julia/src/interpreter.c:608
jl_interpret_toplevel_thunk at /home/keno/julia/src/interpreter.c:692
jl_toplevel_eval_flex at /home/keno/julia/src/toplevel.c:590
jl_parse_eval_all at /home/keno/julia/src/ast.c:907
jl_load at /home/keno/julia/src/toplevel.c:614
jl_load_ at /home/keno/julia/src/toplevel.c:621
include_from_node1 at ./loading.jl:551
unknown function (ip: 0x7fbe34c7069d)
jl_call_fptr_internal at /home/keno/julia/src/julia_internal.h:353
jl_call_method_internal at /home/keno/julia/src/julia_internal.h:372
jl_apply_generic at /home/keno/julia/src/gf.c:1923
include at ./sysimg.jl:14
unknown function (ip: 0x7fbe67acda99)
jl_call_fptr_internal at /home/keno/julia/src/julia_internal.h:353
jl_call_method_internal at /home/keno/julia/src/julia_internal.h:372
jl_apply_generic at /home/keno/julia/src/gf.c:1923
do_call at /home/keno/julia/src/interpreter.c:75
eval at /home/keno/julia/src/interpreter.c:242
jl_interpret_toplevel_expr at /home/keno/julia/src/interpreter.c:34
jl_toplevel_eval_flex at /home/keno/julia/src/toplevel.c:575
jl_eval_module_expr at /home/keno/julia/src/toplevel.c:203
jl_toplevel_eval_flex at /home/keno/julia/src/toplevel.c:478
jl_parse_eval_all at /home/keno/julia/src/ast.c:907
jl_load at /home/keno/julia/src/toplevel.c:614
jl_load_ at /home/keno/julia/src/toplevel.c:621
include_from_node1 at ./loading.jl:551
unknown function (ip: 0x7fbe67d287e9)
jl_call_fptr_internal at /home/keno/julia/src/julia_internal.h:353
jl_call_method_internal at /home/keno/julia/src/julia_internal.h:372
jl_apply_generic at /home/keno/julia/src/gf.c:1923
include at ./sysimg.jl:14
unknown function (ip: 0x7fbe67acda99)
jl_call_fptr_internal at /home/keno/julia/src/julia_internal.h:353
jl_call_method_internal at /home/keno/julia/src/julia_internal.h:372
jl_apply_generic at /home/keno/julia/src/gf.c:1923
do_call at /home/keno/julia/src/interpreter.c:75
eval at /home/keno/julia/src/interpreter.c:242
jl_interpret_toplevel_expr at /home/keno/julia/src/interpreter.c:34
jl_toplevel_eval_flex at /home/keno/julia/src/toplevel.c:575
jl_toplevel_eval at /home/keno/julia/src/toplevel.c:598
jl_toplevel_eval_in at /home/keno/julia/src/builtins.c:496
anonymous at ./<missing>:2
jl_call_fptr_internal at /home/keno/julia/src/julia_internal.h:353
jl_call_method_internal at /home/keno/julia/src/julia_internal.h:372
jl_toplevel_eval_flex at /home/keno/julia/src/toplevel.c:587
jl_toplevel_eval at /home/keno/julia/src/toplevel.c:598
jl_toplevel_eval_in at /home/keno/julia/src/builtins.c:496
eval at ./boot.jl:236
unknown function (ip: 0x7fbe67cbec2d)
jl_call_fptr_internal at /home/keno/julia/src/julia_internal.h:353
jl_call_method_internal at /home/keno/julia/src/julia_internal.h:372
jl_apply_generic at /home/keno/julia/src/gf.c:1923
process_options at ./client.jl:288
_start at ./client.jl:373
unknown function (ip: 0x7fbe67d3bd31)
jl_call_fptr_internal at /home/keno/julia/src/julia_internal.h:353
jl_call_method_internal at /home/keno/julia/src/julia_internal.h:372
jl_apply_generic at /home/keno/julia/src/gf.c:1923
jl_apply at /home/keno/julia/ui/../src/julia.h:1424
true_main at /home/keno/julia/ui/repl.c:127
main at /home/keno/julia/ui/repl.c:264
```
Happens at with ForwardDiff `0.4.2` as well as with `0.5.0`.
```
(rr) p jl_(sig)
Tuple{typeof(Base.convert), Type{ForwardDiff.Dual{N, T} where T<:Real where N}, ForwardDiff.Dual{N, T} where T<:Real where N}
$1 = void
(rr) p jl_(type)
Tuple{typeof(Base.convert), Type{D}, D} where D<:(ForwardDiff.Dual{N, T} where T<:Real where N)
$2 = void
```

**Comment by JeffBezanson** [comment]
I'm instead running into something that looks similar to #22158 here.

**Comment by JeffBezanson** [comment]
What branch/commit of julia is this?

**Comment by Keno** [comment]
88c21f4

**Comment by JeffBezanson** [comment]
I'm having trouble reproducing this. Is there anything else loaded in your session? Are there local modifications to anything?

**Comment by Keno** [comment]
No and no. Are you running with assertions?

**Comment by JeffBezanson** [comment]
Yes; I got that world age assertion. When I commented that out, I got nothing. Do you have any packages on master? What version of DiffBase? ForwardDiff allows 0.2.0 and 0.3.0, but for some reason I can only install 0.2.0.

**Comment by tkelman** [comment]
Try in a clean `JULIA_PKGDIR` if you've got other packages holding things back

**Comment by JeffBezanson** [comment]
Ok, reproduced with these package versions:

```
3 required packages:
 - Compat                        0.25.0             pinned.e628eb0c.tmp
 - ForwardDiff                   0.4.2              pinned.82dcd989.tmp
 - NaNMath                       0.2.4              pinned.cbd6063e.tmp
```

**Comment by JeffBezanson** [comment]
I have a fix for this in #22162 now. Fixing it will involve changing specificity anyway, so if we're ok with that we can eventually backport #22162 to 0.6. If we're not ok with that we can maybe just leave this alone since newer versions of the packages don't seem to exhibit the problem.
